qutebrowser/misc/sql.py

```diff
@@ -475,3 +476,87 @@ def select(self, sort_by, sort_order, limit=-1):
                           sort_order=sort_order))
         q.run(limit=limit)
         return q
+
+
+class Database:
+
+    _database: Optional[QSqlDatabase] = None
+    _user_version: Optional[UserVersion] = None
+    _USER_VERSION = UserVersion(0, 4)  # The current / newest user version
+
+    def __init__(self, db_path: str, connection_name: Optional[str]=None):
+        self._database = QSqlDatabase.database(connection_name, open=False)
+        if not self._database.isValid():
+            self._database = QSqlDatabase.addDatabase('QSQLITE', connection_name)
+            self._database.setDatabaseName(db_path)
+        elif self._database.databaseName() != db_path:
+            # What now? Maybe raise an error? Or we could use db_path both as the
+            # database name **and** the connection name to avoid discrepancies
+            pass
+        if not self._database.isValid():
+            raise KnownError('Failed to add database. Are sqlite and Qt sqlite '
+                             'support installed?')
+        if not self._database.open():
+            error = self._database.lastError()
+            msg = "Failed to open sqlite database at {}: {}".format(db_path,
+                                                                    error.text())
```

Comment:
While you're at it, could you use an f-string for the formatting here?

Reply:
Absolutely